### PR TITLE
Handle prerelease version nuances

### PR DESCRIPTION
**Pull request by snyamathi**
@heikomat this should handle your case along with a couple of other potential issues with prerelease versions.  There's a lot of corner cases here, so let me know if there is anything else tha should be added as a test case.

### CODE REVIEW

File: src/semver-intersect.js [hunk 1]
```diff
@@ -105,11 +108,29 @@ function intersect (...ranges) {
...
+function checkPrereleaseVersions (range, bound) {
+    const parsedRange = parseRange(range);
+    const parsedBound = parseRange(bound);
+
+    if (parsedRange.prerelease && parsedBound.prerelease) {
+        if (
+            (parsedRange.major !== parsedBound.major) ||
+            (parsedRange.minor !== parsedBound.minor) ||
+            (parsedRange.patch !== parsedBound.patch) ||
+            (parsedRange.prerelease[0] !== parsedBound.prerelease[0])
```
Comment (heikomat) [inline src/semver-intersect.js]: The prerelease-tag doesn't have to be identical for the range to be vaild, but **a** prerelease tag has to exist:

bildschirmfoto 2018-02-01 um 14 24 13

> If a version has a prerelease tag (for example, 1.2.3-alpha.3) then it will only be allowed to satisfy comparator sets if at least one comparator with **the same [major, minor, patch] tuple** also has **a prerelease tag**.

File: tests/unit/semver-intersect.js [hunk 2]
```diff
@@ -138,6 +138,30 @@ describe('intersect', () => {
...
+        const result = intersect('^0.14.0-beta', '^0.14.0-beta4');
+        expect(result).to.equal('^0.14.0-beta4');
+    });
+    it('should handle incompatible pre-release versions (different patch version)', () => {
+        const call = intersect.bind(null, '^1.9.0-alpha', '^1.9.1-alpha');
+        expect(call).to.throw('Prerelease version >=1.9.1-alpha is not compatible with >=1.9.0-alpha');
+    });
+    it('should handle incompatible pre-release versions (different preid)', () => {
+        const call = intersect.bind(null, '^1.9.0-alpha', '^1.9.0-beta');
+        expect(call).to.throw('Prerelease version >=1.9.0-beta is not compatible with >=1.9.0-alpha');
```
Comment (heikomat) [inline tests/unit/semver-intersect.js]: That is not correct. `1.9.0-beta` **does** satisfy `^1.9.0-alpha` for the same reason that i mentioned above:

bildschirmfoto 2018-02-01 um 14 27 18

**Review by heikomat — CHANGES_REQUESTED** [review]

**Comment by heikomat** [comment]
@snyamathi Thanks for tackling the issue that fast! i did find something in this PR that contradicts semver. When comparing, semver doesn't seem to care **what** prerelease-tags are used, but only **if** prerelease-tags are used.

File: src/semver-intersect.js (hunk 1 above)
Comment (snyamathi) [reply src/semver-intersect.js]: Gotcha, actually it's one more bit above and beyond that - if the `preId` is the same for both then you also have to look at the number associated with it:

```
> s.satisfies('1.0.0-rc4', '^1.0.0-pre3 <2.0.0')
true
> s.satisfies('1.0.0-rc4', '^1.0.0-pre5 <2.0.0')
true
> s.satisfies('1.0.0-rc4', '^1.0.0-rc5 <2.0.0')
false
```

Comment (snyamathi) [reply src/semver-intersect.js]: > The prerelease-tag doesn't have to be identical for the range to be vaild, but a prerelease tag has to exist

So one more weird thing - it's not that the prerelease tag hast to exist, but it *also* takes alphabetical order into account!  rc4 satisfies ^pre but not the other way around.  Crazy

```
> s.satisfies('1.0.0-alpha', '>=1.0.0-beta <2.0.0')
false
> s.satisfies('1.0.0-beta', '>=1.0.0-alpha <2.0.0')
true
> s.satisfies('1.0.0-beta', '>=1.0.0-betb <2.0.0')
false
> s.satisfies('1.0.0-betc', '>=1.0.0-betb <2.0.0')
true
```

**Comment by snyamathi** [comment]
@heikomat Please task another look at this - I added a bunch of test cases and leveraged semver a bit more to do checking which cleans up things a bit.

File: tests/unit/semver-intersect.js [hunk 3]
```diff
@@ -138,6 +138,39 @@ describe('intersect', () => {
...
+        expect(result).to.equal('^0.14.0-beta4');
+    });
+    it('should handle incompatible pre-release versions (different patch version)', () => {
+        const call = intersect.bind(null, '^1.9.0-alpha', '^1.9.1-alpha');
+        expect(call).to.throw('Range >=1.9.1-alpha is not compatible with >=1.9.0-alpha');
+    });
+    it('should handle compatible pre-release versions (preid lower in alphabetical order)', () => {
+        expect(intersect('^1.9.0-alpha', '^1.9.0-beta')).to.equal('^1.9.0-beta');
+        expect(intersect('^1.9.0-beta', '^1.9.0-alpha')).to.equal('^1.9.0-beta');
+        expect(intersect('^1.9.0-alpha.1', '^1.9.0-beta.2')).to.equal('^1.9.0-beta.2');
```
Comment (snyamathi) [inline tests/unit/semver-intersect.js:164]: These were the test cases that were previously ( erroneously ) failure cases

**Review by heikomat — APPROVED** [review]
If the tests you wrote pass, then this looks good to me! :)

**Comment by snyamathi** [comment]
Sweet - yeah there's CI tests using Travis https://travis-ci.org/snyamathi/semver-intersect/builds/336263754?utm_source=github_status&utm_medium=notification
